Update disk format related function

XueqiangWei opened the pull request:
ID: 1422740 1561839
Signed-off-by: Xueqiang Wei <xuwei@redhat.com>

XueqiangWei commented:
(1/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_
net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.one_disk_repeat: PA
SS (127.22 s)
 (2/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_
net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.max_disk: PASS (414
.49 s)
 (3/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_
net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.all_drive_format_types: PASS (68.08 s)
 (4/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.block: PASS (74.22 s)
 (5/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic: PASS (78.43 s)
 (6/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun: PASS (796.91 s)
 (7/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun: PASS (2446.69 s)
(8/8) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun: PASS (411.78 s)

XueqiangWei commented:
(1/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-g
ithub-autotest-qemu.multi_disk.one_disk_repeat.q35: PASS (223.98 s)
 (2/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.max_disk.q35: PASS (638.53 s)
 (3/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.all_drive_format_types.q35: PASS (180.80 s)
 (4/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.block.q35: PASS (264.08 s)
 (5/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic.q35: PASS (236.31 s)
 (6/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun.q35: PASS (1543.71 s)
 (7/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun.q35: PASS (4193.61 s)
 (8/8) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun.q35: PASS (1059.75 s)

hereischen commented on virttest/utils_disk.py:
```diff
@@ -182,6 +183,585 @@ def clean_old_image(image):
+    :param size: partition size. e.g. 200M
+    :param n_partitions: the number of partitions on disk
+    """
+    partition_size = float(size[0:-1]) / n_partitions
```
Better have `from __future__ import division`

XueqiangWei replied:
size is string, get from normalize_data_size()

hereischen commented on virttest/utils_disk.py:
```diff
@@ -182,6 +183,585 @@ def clean_old_image(image):
+    :param session: session object to guest.
+    :param partition_name: partition name. e.g. sdb1
+    """
+    kname = partition_name[0:-1]
```
`kname = partition_name[:-1]` would be enough

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +183,585 @@ def clean_old_image(image):
+    return mountpoint
+
+
+def format_linux_disk(session, did, size, all_disks_did=None,
```
I think we don't need `all_disks_did`

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +183,585 @@ def clean_old_image(image):
+    :param cmd: cmd to be wrapped.
+    :return: wrapped cmd
+    """
+    cmd_header = "echo list disk > disk && echo select disk %s >> disk "
```
It is not thread-safe: all threads would use the same `disk` cfg file.

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +183,585 @@ def clean_old_image(image):
+    session.cmd_output_safe(mklabel_cmd)
+
+
+def create_partition_table_windows(session, did, labeltype="mbr"):
```
mbr is default value, no need any operation

XueqiangWei commented on virttest/utils_numeric.py:
```diff
@@ -11,3 +14,68 @@ def align_value(value, factor=1024):
+        return ("%.20f" % data_size).rstrip('0')
+
+
+def generate_random_string(length, ignore_str=string.punctuation,
```
this function references to utils_misc.py

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+    create_partition_table_windows(session, did, labeltype)
+    create_partition_batch_windows(session, did, size, n_partitions, labeltype)
+    for i in range(n_partitions):
+        mountpoint.append(set_drive_letter(session, did, partition_no=i+1))
```
If a partition failed to get a drive letter, we should stop to configure the next partition and report the error immediately.

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+        create_partition_table_linux(session, did, labeltype)
+
+
+def create_partition_linux(session, did, start, end, part_type=PARTITION_TYPE_PRIMARY):
```
parted tool usage:
parted mkpart part-type [fs-type] start end

XueqiangWei replied:
updated

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+    details = session.cmd_output(list_partition_cmd % did)
+    for line in details.splitlines():
+        if re.search("Reserved", line, re.I | re.M):
+            partition_no += int(line.split()[1])
```
if create partition table to gpt in windows os, a Reserved partition is generated automatically.

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+    session.cmd(remove_cmd % drive_letter)
+
+
+def format_windows_disk(session, did, size, n_partitions=1, fstype="ntfs",
```
I'd suggest using the function name `configure_empty_$ostype_disk()`.

XueqiangWei replied:
updated

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+    session.cmd(remove_cmd % drive_letter)
+
+
+def format_windows_disk(session, did, size, n_partitions=1, fstype="ntfs",
```
I'm not sure how possible we will use the `size`.

XueqiangWei replied:
the size is get from cfg file, it is a necessary parameter.

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+                create_partition_linux(session, did, str(start) + size[-1],
+                                       str(end) + size[-1], part_type)
+        start, end = end, end + partition_size
+        create_filesyetem_linux(session, did + str(i), fstype)
```
The partition name not always be `did + str(i)` (see my previous comment), so it is better to get the name by `create_partition()`.

XueqiangWei replied:
in order to have the same behavior as windows one, updated it in function \"configure_empty_linux_disk\".

pre_partition = get_linux_disks(session, partition=True).keys()
post_partition = get_linux_disks(session, partition=True).keys()
new_partition = list(set(post_partition) - set(pre_partition))[0]

luckyh commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+        create_filesyetem_linux(session, did + str(i), fstype)
+        mount_dst = "/mnt/" + did + str(i)
+        session.cmd("rm -rf %s; mkdir %s" % (mount_dst, mount_dst))
+        mount("/dev/%s" % did + str(i), mount_dst, fstype=fstype, session=session)
```
Should check if the status of `mount()`.

XueqiangWei replied:
updated

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+    :param session: session object to guest.
+    :param partition_name: partition name. e.g. sdb1
+    """
+    kname = partition_name[:-1]
```
add \"get_kname_cmd = \"lsblk -no pkname /dev/partition_name\"\" to get kname directly

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+    return " && ".join([cmd_header, cmd, cmd_footer])
+
+
+def update_windows_disk_attributes(session, dids, timeout=120):
```
update doc str

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+    output = session.cmd_output(list_disk_cmd)
+    output = output.splitlines()
+    rm_cmd = 'parted -s "/dev/%s" rm %s'
+    for line in output:
```
updated

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+            break
+    else:
+        logging.error("Not found partition '%s', no "
+                      "need to delete it" % partition_name)
```
updated

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,559 @@ def clean_old_image(image):
+                umount("/dev/%s" % line.split()[0], line.split()[-1], session=session)
+            logging.info("remove partition %s on %s" % (partition[0], did))
+            session.cmd(rm_cmd % (did, partition[0]))
+            session.cmd("partprobe")
```
updated

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -182,6 +188,583 @@ def clean_old_image(image):
+    if labeltype == PARTITION_TABLE_TYPE_MBR and n_partitions > 1:
+        part_type = PARTITION_TYPE_EXTENDED
+    else:
+        part_type = PARTITION_TYPE_PRIMARY
```
updated

luckyh commented:
@pingl2016 request for your review, thanks.

lolyu commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,575 @@ def clean_old_image(image):
+    :return: the disks dict.
+             e.g. {kname: [kname, size, type, serial, wwn]}
+    """
+    list_disk_cmd = "lsblk -o KNAME,SIZE,TYPE"
```
is `lsblk -o kname,size,type,wwn,serial` a better choice here?

lolyu commented on virttest/utils_numeric.py:
```diff
@@ -37,3 +38,40 @@ def format_size_human_readable(value, binary=False, precision='%.2f'):
+    if data_size.is_integer():
+        return "%.1f" % data_size
+    else:
+        return ("%.20f" % data_size).rstrip('0')
```
Is it a better choice to use `decimal.Decimal`?

lolyu commented on virttest/utils_numeric.py:
```diff
@@ -37,3 +38,40 @@ def format_size_human_readable(value, binary=False, precision='%.2f'):
+    if data_size.is_integer():
+        return "%.1f" % data_size
+    else:
+        return ("%.20f" % data_size).rstrip('0')
```
`normalize_data_size('123921890', factor='1000')` will get result as `123.9218900000000047612`.
Redundant scale will lead to precision loss, will it be better to use decimal.Decimal?

XueqiangWei commented:
@pingl2016 Please help review it. Many thanks.

balamuruhans commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,571 @@ def clean_old_image(image):
+
+def create_partition_linux(session, did, size, start, part_type=PARTITION_TYPE_PRIMARY):
+    """
+    Create single partition on disk in linux guest.
```
There are some similar APIs in utils_misc.py for linux/windows w.r.t disk operations similar to that you have implemented.

I would request to review, refactor and place them in utils_disk.py.

**example:**
https://github.com/avocado-framework/avocado-vt/blob/56129a616849882ca0b35a4e6309abe379ae0819/virttest/utils_misc.py#L2960-L2975

XueqiangWei replied:
The related functions that you mentioned in utils_misc.py will be replaced by the new implementation in utils_disk.py. We move format disk related functions from utils_misc.py to utils_disk.py, and update them.

hereischen commented on virttest/utils_disk.py:
```diff
@@ -182,6 +190,571 @@ def clean_old_image(image):
+    :return: wrapped cmd
+    """
+    disk = "disk_" + ''.join(random.sample(string.ascii_letters + string.digits, 4))
+    cmd_header = "echo list disk > " + disk
```
Better add `echo rescan > cmd && echo select disk 1 >> cmd && echo clean >> `
May be add a force create partition option.
```
C:\>echo rescan > cmd && echo select disk 1 >> cmd  && echo create partition primary >> cmd && echo select partition 1 >> cmd && echo assign letter=I >> cmd && echo exit >> cmd && diskpart /s cmd

Microsoft DiskPart version 6.1.7601
Copyright (C) 1999-2008 Microsoft Corporation.
On computer: WIN-M02HSVFMND8

Please wait while DiskPart scans your configuration...

DiskPart has finished scanning your configuration.

Disk 1 is now the selected disk.

Virtual Disk Service error:
There is not enough usable space for this operation.


C:\>echo rescan > cmd && echo select disk 1 >> cmd && echo clean >> cmd && echo create partition primary >> cmd && echo select partition 1 >> cmd && echo assign letter=I >> cmd && echo exit >> cmd && diskpart /s cmd

Microsoft DiskPart version 6.1.7601
Copyright (C) 1999-2008 Microsoft Corporation.
On computer: WIN-M02HSVFMND8

Please wait while DiskPart scans your configuration...

DiskPart has finished scanning your configuration.

Disk 1 is now the selected disk.

DiskPart succeeded in cleaning the disk.

DiskPart succeeded in creating the specified partition.

Partition 1 is now the selected partition.

DiskPart successfully assigned the drive letter or mount point.

Leaving DiskPart..
```

XueqiangWei replied:
Thank you in advance.
1. we can get all  empty disks by function \"get_windows_disks_index\" before format disk to avoid force create partition. 
2. if we force create partition, we can call function \"delete_partition_windows\" or \"clean_partition_windows\" before create partition.
3. suggest to delete partition after test. (e.g. get_windows_disks_index ->  configure_empty_disk -> clean_partition_windows

YongxueHong commented:
LGTM.
Acked by: Yongxue Hong <yhong@redhat.com>

YongxueHong commented:
LGTM.
Acked-by: Yongxue Hong <yhong@redhat.com>

lolyu commented:
Ack.

balamuruhans commented:
@XueqiangWei we discussed to move the disk operation related code from utils_misc.py to utils_disk.py which are more similar.

XueqiangWei commented:
@balamuruhans the related functions in utils_misc.py are covered by this patch, and the related code in utils_misc will be rejected in future.

XueqiangWei commented:
@luckyh please help check it again. Many thanks.

XueqiangWei commented:
@YongxueHong as we talked, updated. Please help review it again. Many thanks.  
By the way, \'size = utils_numeric.normalize_data_size(size, order_magnitude=\"M\") \' is also added to create_partition_linux. Because parted tool is not support the size < 1 (e.g. size=0.5G)

XueqiangWei commented:
(01/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.one_disk_repeat: PASS (100.04 s)
 (02/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.max_disk: PASS (202.41 s)
 (03/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.all_drive_format_types: PASS (62.31 s)
 (04/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.block: PASS (100.97 s)
 (05/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic: PASS (102.54 s)
 (06/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun: PASS (1162.07 
s)
 (07/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun: PASS (3
691.79 s)
 (08/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun: PASS (571.94 s)
(09/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.one_disk_repeat: PASS (238.92 s)
 (10/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.max_disk: PASS (638.10 s)
 (11/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.all_drive_format_types: PASS (188.34 s)
 (12/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.block: PASS (277.96 s)
 (13/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic: PASS (231.28 s)
 (14/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun: PASS (1502.97 s)
 (15/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun: PASS (4205.35 s)
 (16/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun: PASS (1018.45 s)

XueqiangWei commented:
@YongxueHong as we talked, updated. Please help review it again. Many thanks.

XueqiangWei commented:
(01/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.one_disk_repeat: PASS (103.44 s)
 (02/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.max_disk: PASS (543.94 s)
 (03/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.all_drive_format_types: PASS (412.40 s)
 (04/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.block: PASS (801.09 s)
 (05/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic: PASS (454.10 s)
 (06/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun: PASS (1163.86 
s)
(07/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.i
o-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun: PASS (3687.60 s)
 (08/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun: PASS (572.89 s)
 (09/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.one_disk_repeat: PASS (575.11 s)
 (10/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.max_disk: PASS (560.46 s)
 (11/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.all_drive_format_types: PASS (176.98 s)
 (12/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.block: PASS (234.30 s)
 (13/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.generic: PASS (237.05 s)
 (14/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_lun: PASS (1532.27 s)
 (15/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_scsiid_lun: PASS (4318.77 s)
(16/16) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.multi_disk.virtio_scsi_variants.multi_bus_scsiid_lun: PASS (993.47 s)

YongxueHong commented:
@XueqiangWei , @luckyh 
Acked-by: Yongxue Hong <yhong@redhat.com>

luckyh commented:
Thanks all.

lolyu commented on virttest/utils_numeric.py:
```diff
@@ -37,3 +40,38 @@ def format_size_human_readable(value, binary=False, precision='%.2f'):
+        raise ValueError("Invalid data size format 'value_str=%s'" % value_str)
+    from_index = __get_unit_index(unit)
+    to_index = __get_unit_index(order_magnitude)
+    scale = int(factor) ** (to_index - from_index)
```
`scale = Decimal(factor) ** (to_index - from_index)` will solve the precision error.

YongxueHong replied:
>>> from virttest import utils_numeric
>>> utils_numeric.normalize_data_size('1M', 'K', '1000')
'1E+3'
>>> 
From the above result , may be it's not a friendly readable format.

lolyu replied:
could disable `Decimal` scientific  representation just like:
`return format(data_size, 'f')`
float type has the same implicit scientific problem, so it is better to add extra parameter to let user to pass a format string like '%.2f', using format string could easily convert into string without scientific notation.
